### ISSUE TITLE
Provide some way to figure out if a battery is not fully charged (without requring at least a 10% charge rate).

### ISSUE THREAD

**Issue by Orum** [Regalis11/Barotrauma#8481]
This sounds like it should be a simple task.  A naive approach would be to simply compare `CHARGE_%` on a battery output to 100.  However, as only integer values are received in output, a battery that is `> 99.5%` charged will read out as `100%`, despite not being fully charged.  A better approach would be to compare the `CHARGE_OUT` to the maximum charge level of a battery.  However, this also suffers the same issue; a battery that is `> 1999.5` charge will read out `2000`, leading circuitry to believe the battery is fully charged when it is not (assuming 2000 is the maximum charge).

Another approach would be using junction boxes with their `LOAD_VALUE_OUT` connected to the battery's `POWER_IN`, but this requires the `CHARGE_RATE` of the battery to be at least `10%`.  Unfortunately this also will not work for many scenarios (including the one I currently find myself in) as I need to know if a battery is not fully charged *even if* its current charge is `1999.9` and its charge rate is currently `0%`.

I propose three decent ways of fixing this, each with trade-offs:
- **The compatible fix:** Add a new output on batteries that is just a `1` if they are fully charged or a `0` if they are not.  While fully compatible for all existing submarines, it adds UI clutter and potentially requires running more wires.
- **The simple fix:** Simply run the internal floating point value through floor() before passing it out for `CHARGE_%` or `CHARGE_OUT`.  This maintains compatibility with anything that expects only an integer value, but has the slight downside that batteries that are nearly empty may appear as fully empty.
- **The powerful fix:** Instead of converting `CHARGE_%`/`CHARGE_OUT` to integers, just pass out their floating point values, and let people do what they want with them (as round, floor, and ceiling components already exist in the game).  Potentially causes issues with things that expect only integer values as input, but gives submarine creators/engineers the most flexibility.

Personally I'd prefer the powerful fix, but I can see the appeal in the simple fix if greater backwards compatibility is required.

**Comment by cshingle** [comment]
Doing a floor would be a simple change but I don't see this as a big issue.
If the battery is %99.5 or %100 charged is a difference of 10 kWmin with the default batteries. That is 1/10th of the efficiency loss added in the new version. I don't see how that is going to make or break a sub.

**Comment by Orum** [comment]
It might not seem like a big difference, but consider the situation I'm in.  I have a charge controller that's controlling the charge in 4 batteries on my submarine, where it selects the charge rate of each battery based on the formula: `(MaxReactorOutput - NonBatteryLoad) / (MaxBatteryChargeRate * BatteriesRequestingCharge)`.  It's actually slightly more complicated than this, as I need to convert to percent, force it to round down to the nearest 10%, prevent division by zero, etc., but it 's set up so that battery charging will never pull more from the reactor than it can provide, assuming it has valid inputs for all of those.

The batteries use the default values in the editor, and my reactor outputs a maximum of 4000W (assuming no upgrades).  Now, let's say I'm in one of two different situations.  The first is that I have one battery that's actually drained, and all others are in fact full.  This can happen if a battery reached 0 condition while empty, and the other three recharged fully while it was still there.  If that battery is then repaired, it now needs to be charged.  Say I have a 3500W non-battery load, so we get: `(4000 - 3500) / (500 * 1) = 500 / 500 = 100% = 500W load`, which is correct, and it brings the reactor up to the max load of 4000W.

Now consider a *nearly identical* situation, except those other three batteries aren't *actually* full, they just appear to be due to the rounding.  Now the batteries are attempting to draw 2000W instead of the 500W the charge controller expects.  This gets into a self-sustaining cycle, where a brown-out is caused as all the batteries are trying to charge at a rate faster than they should, which causes the batteries to output (as they're trying to cover for the brown-out), which keeps them slightly discharged and thus still charging (without the controller knowing they are charging), continuing the brown-out, and keeping the cycle going.  Then, to add insult to injury, the batteries themselves consume some power as they're only 95% efficient, so the load is even higher than if no "full but not actually full" batteries were in the mix.

**Comment by cshingle** [comment]
Your circuit will not work how you think it should. The issue is that if there is any load on the batteries, even a single lamp they will likely never reach a %100 charge. Removing the rounding actually makes it impossible for the battery to reach 100 charge when there is a load. (I actually tested this)
The reason for this is that the battery receives power from a source (The reactor) and in the same tick another consumer will pull from it, so by the end of the tick is is no longer %100. Without the rounding this becomes really obvious because it will charge to %99 and just stay there until you remove the load.
I strongly recommend modifying your charge controller to reduce the charge rate as the battery approaches %100. A lot of custom subs do this to address this exact issue.
This will actually be less of a problem with the current unstable is released since it adds prioritization of power sources rather than the round robin that the current stable uses. The prioritization will make devices preferentially draw power from the reactor and will only pull from batteries wired into the same grid if the reactor can't provide enough. This basically emulates the power system in Factorio which is excellent.

**Comment by Orum** [comment]
The batteries are not outputting 100% of the time.

The output on the batteries is entirely dependent on the load on the reactor vs what the reactor is actually outputting.  If the `SuppliedPower / TotalLoad < 0.9`, that is, if the reactor is fulfilling less than 90% of the load, I consider this a brown-out and turn the battery output on.  The problem is that because the charge controller thinks the batteries are fully charged, it's actually *creating* the load that causes the batteries to think there is a brown-out.

I haven't tried the unstable builds yet as I have enough bugs with the stable.

**Comment by cshingle** [comment]
> I haven't tried the unstable builds yet as I have enough bugs with the stable.

That is understandable. Unstable is pretty unstable right now.
If you have a relay on the power output side then the batteries will charge and stop pulling load on the input side. If the output is wired to the grid then the batteries will always have passthrough load. The current power system pulls from every power source that is connected to the grid. That is why batteries discharge even if the reactor has more than enough head room to handle the load.

**Comment by Orum** [comment]
> If you have a relay on the power output side then the batteries will charge and stop pulling load on the input side.
Yes.  The issue is once they *ever* output *any* amount of charge, but the charge controller thinks they are charged, they will *never* fully be charged (as you pointed out in your earlier post).  Thus, once the cycle starts, it will never stops until the total load is less than the reactor can output.
